src/placeos-driver/settings/introspect.cr

```diff
@@ -0,0 +1,137 @@
+    end
+  end
+
+  module ::JSON::Serializable
+    macro included
+      extend ::PlaceOS::Introspect
+    end
+  end
+
+  macro introspect(klass)
+    {% arg_name = klass.stringify %}
+    {% if !arg_name.starts_with?("Union") && arg_name.includes?("|") %}
+      PlaceOS.introspect(Union({{klass}}))
+    {% else %}
+      {% klass = klass.resolve %}
+      {% klass_name = klass.name(generic_args: false) %}
+
+      {% if klass <= Array %}
+        {% if klass.type_vars.size == 1 %}
+          has_items = PlaceOS.introspect {{klass.type_vars[0]}}
+        {% else %}
+          has_items = {} of String => String
+        {% end %}
+        if has_items.empty?
+          %klass = {{klass}}
+          %klass.responds_to?(:json_schema) ? %klass.json_schema : {type: "array"}
+        else
+          {type: "array", items: has_items}
+        end
+      {% elsif klass.union? %}
+        { anyOf: [
+          {% for type in klass.union_types %}
+            PlaceOS.introspect({{type}}),
+          {% end %}
+        ]}
+      {% elsif klass_name.starts_with? "Tuple(" %}
+        has_items = [
+          {% for generic in klass.type_vars %}
+            PlaceOS.introspect({{generic}}),
+          {% end %}
+        ]
+        {type: "array", items: has_items}
+      {% elsif klass_name.starts_with? "NamedTuple(" %}
+        {type: "object",  properties: {
+          {% for key in klass.keys %}
+            {{key.id}}: PlaceOS.introspect({{klass[key].resolve.name}}),
+          {% end %}
+        },
+          {% required = [] of String %}
+          {% for key in klass.keys %}
+            {% if !klass[key].resolve.nilable? %}
+              {% required << key.id.stringify %}
+            {% end %}
+          {% end %}
+          {% if !required.empty? %}
+            required: [
+            {% for key in required %}
+              {{key}},
+            {% end %}
+            ]
+          {% end %}
+        }
+      {% elsif klass < Enum %}
+        {type: "string",  enum: {{klass.constants.map(&.stringify)}} }
+      {% elsif klass <= String %}
+        { type: "string" }
+      {% elsif klass <= Bool %}
+        { type: "boolean" }
+      {% elsif klass <= Int %}
+        { type: "integer" }
+      {% elsif klass <= Float %}
+        { type: "number" }
+      {% elsif klass <= Nil %}
+        { type: "null" }
+      {% elsif klass <= Hash %}
+        {% if klass.type_vars.size == 2 %}
+          { type: "object", additionalProperties: PlaceOS.introspect({{klass.type_vars[1]}}) }
+        {% else %}
+          # As inheritance might include the type_vars it's hard to work them out
+          %klass = {{klass}}
+          %klass.responds_to?(:json_schema) ? %klass.json_schema : { type: "object" }
+        {% end %}
+      {% elsif klass.ancestors.includes? JSON::Serializable %}
+        {{klass}}.__generate_json_schema__
+      {% else %}
+        %klass = {{klass}}
+        if %klass.responds_to?(:json_schema)
+          %klass.json_schema
```

Comment:
absolutely not required at this stage but it would be good if this is used extensively to have some json schema validation capability on the return value of `def json_schema` to make sure its following json schema specs (eg raise compiler error if key value is false, e.g. `require` when it should be `required`. this [lib](https://github.com/sa-0001/json-schema-cr) doesn't seem too clean but I think we can use some of the idea if this is to be implemented in the future.